gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GCSManualClientImpl.java

```diff
@@ -0,0 +1,222 @@
+package com.google.cloud.hadoop.gcsio;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.auth.Credentials;
+import com.google.cloud.hadoop.util.AccessBoundary;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import java.io.IOException;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Provides read/write access to Google Cloud Storage (GCS), using Java nio channel semantics. This
+ * is a basic implementation of the GoogleCloudStorage interface that mostly delegates through to
+ * the appropriate API call(s) google-cloud-storage client.
+ */
+public class GCSManualClientImpl implements GoogleCloudStorage {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  /**
+   * Having an instance of gscImpl to redirect calls to Json client while new client implementation
+   * is in WIP.
+   */
+  private GoogleCloudStorageImpl gcsClientDelegate;
+
+  private GoogleCloudStorageOptions storageOptions;
+  private Credentials credentials;
+
+  public GCSManualClientImpl(GCSManualClientImplBuilder builder) throws IOException {
+    this.storageOptions = checkNotNull(builder.storageOptions, "options must not be null");
```

Comment:
these checks can go to Builder.build() method.

Reply:
I am delegating this check to the constructor as it has all info to figure out for what all fields null check is required.
Although I agree in this case it wouldn't matter much but I find this null checks better to be placed in constructor.